### PR TITLE
feat(auth): Implement Passwordless Authentication

### LINKED ISSUE #8165: Move email login to passwordless
The first part of simplifying our authentication issues is to move to passwordless login. 

~~For this we should use some sort of emailed-out token, such as this library:
https://github.com/florianheinemann/passwordless~~

This involves: 
- [x] re-add email verification
- [x] create user email sign up logic
- [x] change email login to use temp tokens sent via email to login
- [x] removing password reset logic
- [x] restrict API access to the user creation

DevOps
- [ ] Remove all passwords from database (@berkeleyTrue)

Edit(@berkeleyTrue):
Removed passwordless module. This is well within the abilities of the loopback framework

### LINKED ISSUE #9027: Confirming your email should also automatically sign you in
Currently, when you click the link to verify your email, it takes you back to FCC and shows you this, but it doesn't log you in.

### LINKED ISSUE #15254: [beta] Sign in issue with username
#### Issue Description

I'm trying to sign back in to the beta site for Free Code Camp, but it's not letting me use my username to login.

It won't let me sign in unless I use an email, even if it shows inside the message box as "Invalid username".  First image is me trying to use my username. Second image is what shows up in the error message box.

#### Browser Information

* Browser Name: Chromium. Version: 59.0.3042.0 (Developer Build) (32-bit)
* Operating System: Windows 8
* Mobile, Desktop, or Tablet: Desktop

#### Screenshot

fcc_issue2

fcc_issue

### PR DESCRIPTION
**Pull request by raisedadead**
Feature Checklist:
- [x] Create API for creating/saving a temporary token
- [x] Add email template and sending logic
- [x] Add validations and timeouts
- [x] Update the sign in view
- [x] Prevent users without verified email from login
- [x] ~~Create API for creating/saving an access token~~ Using loopbacks defaults
- [x] Add routes for requesting access token
- [x] Invalidate used/expired tokens
- [x] Remove password reset logic
- [x] Handle all errors gracefully
- [x] Remove all passwords from database (#14559 issue created for DevOps)
- [x] Add debounce/disable to magic link button
- [x] Add some loading indicator to magic link button ajax request
- [ ] (berks): Investigate memory consumption issue

Closes #8165 
Closes #9027
Closes #15254

### CODE REVIEW

**Comment by QuincyLarson** [comment]
@atjonathan @raisedadead @BerkeleyTrue I have created a new issue related to this here: https://github.com/FreeCodeCamp/FreeCodeCamp/issues/10407

**Comment by BerkeleyTrue** [comment]
@raisedadead Please continue to work on this. #10407 does not change the user stories of this PR.

**Comment by QuincyLarson** [comment]
Reopening this because the underlying passwordless functionality is still needed for updating the workflow.

**Comment by QuincyLarson** [comment]
@raisedadead I know you're recovering from the flu. I just wanted to check and see how this is coming along, and let you know that going passwordless is still a priority for us 😄

**Comment by raisedadead** [comment]
Hi @quincylarson, just got back. I plan to work on this from tomorrow. Sorry about the delay.

**Comment by QuincyLarson** [comment]
@raisedadead no need to apologize - thanks for helping us implement this!

**Comment by raisedadead** [comment]
Just a rebase of the mess that I created. Will be ready with the feature in the next push.

**Comment by BerkeleyTrue** [comment]
👍

**Comment by raisedadead** [comment]
Basic functionality is there out now… just some more validation touches and code optimisations and we are good (soon).

**Comment by QuincyLarson** [comment]
@raisedadead awesome - thank you for keeping us posted 😄

**Comment by raisedadead** [comment]
This is mostly ready, but am at my wits end for getting the request to complete when the link in the email is completed.

Basically I need to get the middleware `invalidateAuthToken` function to execute correctly.

Would appriciate a review and guidance from anyone.

**Comment by raisedadead** [comment]
Never mind my last comment. I figured out that I was just more sleepy than I should have been last night.

This is ready for QA.

File: common/models/user.js
Comment (BerkeleyTrue) [inline common/models/user.js]: Instead of creating this renderer per request, can we move this up into the module scope and then use the ternary above to select the appropriate renderer instead of just the template name?

File: common/models/user.js
Comment (BerkeleyTrue) [inline common/models/user.js]: Async errors must be handled correctly, throwing here will cause servers to crash unexpectedly.

File: common/models/user.js
Comment (BerkeleyTrue) [inline common/models/user.js]: You must use the update$ methods when modifying the user object.

File: common/models/user.js
Comment (BerkeleyTrue) [inline common/models/user.js]: indenting

File: common/models/user.js
Comment (BerkeleyTrue) [inline common/models/user.js]: indenting

File: common/models/user.js
Comment (BerkeleyTrue) [inline common/models/user.js]: Errors must be dealt with by express here. Do not use a catch here unless modifying the error message and then rethrowing

File: common/models/user.js
Comment (BerkeleyTrue) [inline common/models/user.js]: isDev shouldn't be used like this. dev and production must run the same.

File: server/boot/user.js
Comment (BerkeleyTrue) [inline server/boot/user.js]: indenting

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

File: common/models/user.js
Comment (raisedadead) [inline common/models/user.js]: I agree, will look into an alternate way for this. If you remember we did this for returning the error to the Promise (AJAX) from the client.

File: common/models/user.js
Comment (BerkeleyTrue) [inline common/models/user.js]: Errors here will be passed down the pipeline until a catch operator is hit or a the onError handler in the subscribe method is called. The err is not passed into the map operator.

**Comment by raisedadead** [comment]
@BerkeleyTrue @QuincyLarson This is ready for a QA. 
Thanks a ton to Berkeley for taking the time to help me through this.

I have discovered a few issues with the updateEmail API while doing a load test, and will fix that in a separate PR, because its unrelated and we have open issues for them I checked.

Nonetheless this is ready for a review and deployment on beta for a testing. Berkeley I could not reproduce the issue about bad tokens, maybe beta is a good candidate for a try out!

Also I have created a DevOps Issue for deleting the passwords, which Berkeley can take up when this reaches production.

Awaiting feedback! Cheers.

**Comment by augmt** [comment]
@raisedadead In accordance with #10407, I think the sign-in/sign-up page should illustrate to campers that signing up for a freeCodeCamp account with GitHub is not allowed.

https://www.freecodecamp.com/signin currently does this best as it does not force returning campers to click an extra time like https://beta.freecodecamp.com/en/signup does and it separates supported sign-in methods from supported sign-up methods. What do you think?

**Comment by raisedadead** [comment]
@augmt That is correct, and I plan to make a separate PR as soon as we complete testing this one. We are limiting the change in this PR to email auth and updates.

**Review by BerkeleyTrue — APPROVED** [review]

**Comment by BerkeleyTrue** [comment]
At the moment, links sent out by beta in the email do not add the subdomain `beta`. This is not the fault of this PR but we will need to add some sort of feature to include this ability before we can test this on Beta. We can either do it here or as a separate PR.

**Comment by raisedadead** [comment]
@BerkeleyTrue I'll add it in this PR itself as soon as I am back tonight. Thanks for the review.

**Comment by raisedadead** [comment]
Hey @BerkeleyTrue I have created a new utility for future use to compute the URLs including things like HOST, PORT etc.

Its a overkill as of now, but I am sure would help in the long run.

Finally I have made a minor change in a previous commit to this PR ensuring that at least the message for a success case reaches the client while updating email.

Please check and let me know how this looks.

File: server/utils/url-utils.js
Comment (BerkeleyTrue) [inline server/utils/url-utils.js]: nit: use a if/esle condition for readability here. 
fix: if not dev port should be omitted.

File: server/utils/url-utils.js
Comment (BerkeleyTrue) [inline server/utils/url-utils.js]: fix: If not dev this should not include port number or the column.

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Comment by raisedadead** [comment]
Done!

File: server/utils/url-utils.js
Comment (BerkeleyTrue) [inline server/utils/url-utils.js]: fix: should not add port in production. This should be null

File: server/utils/url-utils.js
Comment (BerkeleyTrue) [inline server/utils/url-utils.js]: fix: This should not include port in production

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

File: server/utils/url-utils.js
Comment (raisedadead) [inline server/utils/url-utils.js]: Yup, isn't the production case in the conditional, i.e the previous block?

File: server/utils/url-utils.js
Comment (raisedadead) [inline server/utils/url-utils.js]: Ah, cool.. I'll update, I used this from the previous code for the verify email send logic, where I think the port was used for production. But NVM, I'll update asap.

File: server/utils/url-utils.js
Comment (BerkeleyTrue) [inline server/utils/url-utils.js]: Ah, I see. I was pattern matching with the previous function.

File: server/utils/url-utils.js
Comment (BerkeleyTrue) [inline server/utils/url-utils.js]: Let's swap these two conditionals so they read a little more obviously.

if (isDev) {
}

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Review by BerkeleyTrue — APPROVED** [review]

**Comment by BerkeleyTrue** [comment]
Through this up on beta and there are still a couple of issues: 

* link sent in email is still pointing to freecodecamp and not beta
* If I press the 'get magic link` button several times in a row, it will send me several emails back to back
     * (client side) to prevent this we need to disable the button while there is a request in flight
     * (server side) We need to figure out some way to lock the user object to prevent the user from constantly sending out emails (unchecked this could provide a vulnerability that could crash servers)

**Comment by raisedadead** [comment]
> link sent in email is still pointing to freecodecamp and not beta

I tested by setting the below, could NOT reproduce:

```
NODE_ENV=production
BETA=true
```
The link is correctly generated, for me.

> If I press the 'get magic link` button several times in a row, it will send me several emails back to back

Yup, we need a strong mechanism for this. 

Additionally, I am seeing a completely different issue, where createAccessToken from loopback is not correctly associating the correct userId to the accessToken that it generates.

This is a weird one.

Can you hit me up in chat when you can?

**Comment by raisedadead** [comment]
@BerkeleyTrue 

Regarding the second concern for multiple request, I have two proposals:

1. Disable the submit, button while the AJAX is in progress.
2. Migrate the Sign up page to react and use RxJS debounce.

I'll let you suggest which one would be a better alternative?

**Comment by BerkeleyTrue** [comment]
The problem lies mainly server side. The question is how do you lock a req path by the same user? Like transactions with databases only for expressjs routes.

**Comment by BerkeleyTrue** [comment]
Maybe we can implement some sort of queuing system to send the emails?

**Comment by BerkeleyTrue** [comment]
Looks like the beta flag issue is solved.

**Comment by raisedadead** [comment]
@BerkeleyTrue: 
- Added the disable and loading indication to the button when the AJAX is in flight. 
- For additional debouncing, I think we can take that at the server side rather.

File: server/utils/url-utils.js
Comment (BerkeleyTrue) [inline server/utils/url-utils.js]: ping

**Comment by BerkeleyTrue** [comment]
@raisedadead I think we can merge this as soon as the merge conflicts are done. I'll run through it one more time and then we will through it up on beta

**Comment by raisedadead** [comment]
This is just an initial rebase of the conflicts, which I basically eye-balled. I am yet to check the changes that I would need to make to get this working as it was earlier.  :rotating_light: :hand:

**Comment by BerkeleyTrue** [comment]
@raisedadead Somehow all the refactoring I did in user.js has been undone in this PR

**Comment by raisedadead** [comment]
@BerkeleyTrue yes, that is correct. I might have messed up the rebase a bit. And have been away a while on an outing. Please give a day or so to clean this up correctly. I'll let you know as soon as it ready.

**Comment by BerkeleyTrue** [comment]
@raisedadead ping

**Review by ghost — APPROVED** [review]

**Comment by frutsy007** [comment]
How is work

**Comment by BerkeleyTrue** [comment]
@piotrek1543 @frutsy007 refrain from unrequested PR review approvals and unrelated comments.

**Comment by raisedadead** [comment]
Hey @BerkeleyTrue so sorry about the delay, I have been a swamped a bit lately for the past couple weeks. I plan to work on this tonight.

**Comment by BerkeleyTrue** [comment]
@raisedadead no worries. Thanks for the update. I'm excited to see this merged in!

**Comment by BerkeleyTrue** [comment]
woot!!

**Comment by raisedadead** [comment]
@BerkeleyTrue this is ready for QA. Please check when you can. Thanks.

**Comment by BerkeleyTrue** [comment]
Looks good. I'll merge when I get to a computer

**Comment by BerkeleyTrue** [comment]
merged!!!!!!!!!!!!!!!!!!!!
https://github.com/freeCodeCamp/freeCodeCamp/commit/61d740b8a0fb144e77a6fa963d0ec7fe16cfead4

✨ 🎊 🚀

I owe you a 🍺 

Happy Coding!

**Comment by raisedadead** [comment]
@BerkeleyTrue awesome, thanks a lot for the QA. Do let me know how it goes when deployed? 
Of course, let's grab some 🍺 when we do meet in person in future!  😄

**Pull request closed without merging**
